frame/vesting/src/vesting_info.rs

```diff
@@ -0,0 +1,131 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Module to enforce private fields on `VestingInfo`.
+
+use super::*;
+use codec::MaxEncodedLen;
+
+/// Struct to encode the vesting schedule of an individual account.
+#[derive(Encode, Decode, Copy, Clone, PartialEq, Eq, RuntimeDebug, MaxEncodedLen)]
+pub struct VestingInfo<Balance, BlockNumber> {
+	/// Locked amount at genesis.
+	locked: Balance,
+	/// Amount that gets unlocked every block after `starting_block`.
+	per_block: Balance,
+	/// Starting block for unlocking(vesting).
+	starting_block: BlockNumber,
+}
+
+impl<Balance, BlockNumber> VestingInfo<Balance, BlockNumber>
+where
+	Balance: AtLeast32BitUnsigned + Copy,
+	BlockNumber: AtLeast32BitUnsigned + Copy + Bounded,
+{
+	/// Instantiate a new `VestingInfo`.
+	pub fn new<T: Config>(
+		locked: Balance,
+		per_block: Balance,
+		starting_block: BlockNumber,
+	) -> VestingInfo<Balance, BlockNumber> {
+		VestingInfo { locked, per_block, starting_block }
+	}
+
+	/// Validate parameters for `VestingInfo`. Note that this does not check
+	/// against `MinVestedTransfer` or the current block.
+	pub fn validate<BlockNumberToBalance: Convert<BlockNumber, Balance>, T: Config>(
+		&self,
+	) -> Result<(), Error<T>> {
+		ensure!(!self.locked.is_zero(), Error::<T>::InvalidScheduleParams);
+
+		let max_block = BlockNumberToBalance::convert(BlockNumber::max_value());
+		let starting_block = BlockNumberToBalance::convert(self.starting_block());
+		match self.locked.checked_div(&self.per_block) {
+			None => return Err(Error::<T>::InfiniteSchedule), // `per_block` is 0
+			Some(duration) => {
+				let end = duration.saturating_add(starting_block);
+				ensure!(end < max_block, Error::<T>::InfiniteSchedule)
```

Comment:
This is not really a fair check. If we every reach `BlockNumber::max_value()` the whole blockchain will break.

By that point, there would be a migration updating the type used for block number, for example from u32 to u64

Reply:
But don't we want to keep people from putting absurdly long schedules in?  Its probably a mistake if someone puts a schedule in this long. Its not really about anything breaking its just saying "you probably didn't mean to do this". (Also if my quick mafs it right we have like 800 years with a u32).

Happy to take the check out though if you want.

Also wondering if we should expose a `remove_schedule` extrinsic - no super strong justification other than spam mitigation.

Reply:
5c02c57a84b71190ff304f6f2f70e11232ee7827